### PR TITLE
Fix more bytes/unicode issues in twisted.web

### PR DESCRIPTION
**Pull request by rodrigc**
https://twistedmatrix.com/trac/ticket/9135

### CODE REVIEW

**Comment by rodrigc** [comment]
@markrwiliams please look

File: src/twisted/web/distrib.py
Comment (markrwilliams) [inline src/twisted/web/distrib.py]: This is UTF-8 because name comes from a URL path?

File: src/twisted/web/resource.py
Comment (markrwilliams) [inline src/twisted/web/resource.py]: `getattr` works in terms of native strings, not unicode strings.  Why did you change this?

File: src/twisted/web/script.py
Comment (markrwilliams) [inline src/twisted/web/script.py]: This should use `sys.getfilesystemencoding`, not UTF-8.

File: src/twisted/web/script.py
Comment (markrwilliams) [inline src/twisted/web/script.py]: [RFC 7230 says header values should be ASCII](https://tools.ietf.org/html/rfc7230#page-25).  Why has this been changed to UTF-8?

File: src/twisted/web/script.py
Comment (markrwilliams) [inline src/twisted/web/script.py:188]: The docstring for `PythonScript` should say that it requires the script to be UTF-8.

File: src/twisted/web/static.py
Comment (markrwilliams) [inline src/twisted/web/static.py]: This should use `sys.getfilesystemencoding`.

File: src/twisted/web/static.py
Comment (markrwilliams) [inline src/twisted/web/static.py]: Headers should be ASCII.

File: src/twisted/web/static.py
Comment (markrwilliams) [inline src/twisted/web/static.py]: You need to use the decoded path: https://github.com/twisted/twisted/pull/894#discussion_r143305534

File: src/twisted/web/static.py
Comment (markrwilliams) [inline src/twisted/web/static.py:464]: Header value should be ASCII, and this goes into a header.

File: src/twisted/web/static.py
Comment (markrwilliams) [inline src/twisted/web/static.py]: Header values should be ASCII.

File: src/twisted/web/static.py
Comment (markrwilliams) [inline src/twisted/web/static.py:529]: Header values should be ASCII.

File: src/twisted/web/static.py
Comment (markrwilliams) [inline src/twisted/web/static.py]: Header values should be ASCII.

File: src/twisted/web/static.py
Comment (markrwilliams) [inline src/twisted/web/static.py]: Header values should be ASCII.

File: src/twisted/web/static.py
Comment (markrwilliams) [inline src/twisted/web/static.py]: Header values should be ASCII.  Also, is the `type` instance variable supposed to allow native strings?  Is that why this can be unicode?

File: src/twisted/web/static.py
Comment (markrwilliams) [inline src/twisted/web/static.py:1089]: This can cause an encoding error with UTF-8 on Python 2, because `done` will first be decoded as ASCII.  The `_PY3` check should remain.

File: src/twisted/web/test/test_static.py
Comment (markrwilliams) [inline src/twisted/web/test/test_static.py]: This should use `sys.getfilesystemencoding`

File: src/twisted/web/test/test_static.py
Comment (markrwilliams) [inline src/twisted/web/test/test_static.py]: Header values should be ASCII.

File: src/twisted/web/test/test_static.py
Comment (markrwilliams) [inline src/twisted/web/test/test_static.py:921]: This is good.  It's more explicit than pretending `BytesIO` is `StringIO`.

File: src/twisted/web/test/test_static.py
Comment (markrwilliams) [inline src/twisted/web/test/test_static.py:1861]: This is good.

File: src/twisted/web/test/test_xmlrpc.py
Comment (markrwilliams) [inline src/twisted/web/test/test_xmlrpc.py]: Why is this URL encoded as UTF-8, when the others are encoded as ASCII?

The best thing to do would be to use `hyperlink.URL.from_text`, then `to_url()`, then encode _that_ as ASCII.  That will encode each component of the URL according to its type.  @mahmoud Can you confirm this?

File: src/twisted/web/xmlrpc.py
Comment (markrwilliams) [inline src/twisted/web/xmlrpc.py]: Please move this to a separate PR.

**Review by markrwilliams — CHANGES_REQUESTED** [review]
Thank you for working on this.

There are changes around header values unrelated to either https://twistedmatrix.com/trac/ticket/8985 or https://twistedmatrix.com/trac/ticket/9135, and unrelated improvements to `t.w.xmlrpc`.

Please move the `xmlrpc` work to another PR, and refrain from including unrelated changes in subsequent PRs.  It makes them much harder to review.

File: src/twisted/web/resource.py
Comment (rodrigc) [reply src/twisted/web/resource.py]: `getattr` works with unicode strings on Python 2 and Python 3.  I confirmed by trying it on both.

Comment (markrwilliams) [reply src/twisted/web/resource.py]: On Python 2, `getattr` will silently encode unicode strings with the ASCII codec.  So this will decode `request.method` as ASCII, concatenate it with u`render_`, then re-encode it as ASCII.

On Python 3, prefixing a string literal with `u` has no effect, and `request.method.decode('ascii')` is equivalent to `nativeString`.

So, yes it works, but `getattr`'s actual API works in terms of native strings. The original code aligned with that API by concatenating a native string with `nativeString`.

File: src/twisted/web/script.py
Comment (rodrigc) [reply src/twisted/web/script.py]: copyright.version always has characters which fall under the ascii range when encoding with utf-8.

File: src/twisted/web/script.py
Comment (rodrigc) [reply src/twisted/web/script.py:188]: PEP 3120 says the default encoding is UTF-8, so that is a reasonable default: https://docs.python.org/3/reference/lexical_analysis.html#lexical-analysis

File: src/twisted/web/test/test_static.py
Comment (rodrigc) [reply src/twisted/web/test/test_static.py]: This value will always be in the ascii range even when encodign with "utf-8".

File: src/twisted/web/static.py
Comment (rodrigc) [reply src/twisted/web/static.py]: I found that when I used the decoded path here, it didn't work due to cascading dependencies in the call chain of this function, and I needed to keep it as bytes.

File: src/twisted/web/static.py
Comment (rodrigc) [reply src/twisted/web/static.py:1089]: No, this is always needed on Python 2 and 3.  Now, done is always unicode, due to header being unicode,
but bytes must be returned from this function.

**Comment by rodrigc** [comment]
@markrwilliams I've updated this patch based on your review comments.  Please look.

File: src/twisted/web/resource.py
Comment (jlitzingerdev) [inline src/twisted/web/resource.py:247]: Am I mistaken, if method happens to be unicode, this will raise an AttributeError where before nativeString() would have just done the right thing?

File: src/twisted/web/test/requesthelper.py
Comment (jlitzingerdev) [inline src/twisted/web/test/requesthelper.py:352]: This comment applies to all the `intToBytes` conversions -- is this safe?  Specifically, intToBytes does provide some validation of the input where these changes will not.  For example, with trunk on both python2 and 3, the following:

```
from twisted.web.http import HTTPChannel, Request

ch = HTTPChannel()
r = Request(ch)
r.setHost(b'http://www.example.com', 'abc')
```
Will raise a TypeError, however with these changes, no error will be raised.

**Review by jlitzingerdev — CHANGES_REQUESTED** [review]
Here's my review by request.  There is a ton of work here, nicely done!

It also seems like the scope of changes is far wider than just fixing putChild(), are there other related tickes or is it general Python3 cleanup?  Specifically, a theme of these changes seems to be to eliminate the use of several (nativeString, networkString, intToBytes) compat functions

My concerns/requested changes, in summary (with details in the review):

Requested Changes:
misc newsfragments don't really seem appropriate.  Even at the most basic level this modifies putChild's behavior in such a way that a client whose code silently did the wrong thing on python 3 will now raise a TypeError.  It would be nice if this were called out in the changelog.

Concerns
1.  I'm worried about the intToBytes removals, specifically the new code is a lot more forgiving than intToBytes.
2.  The scope of changes also seems to include a general move away from the twisted.python.compat functions (nativeString, networkString, intToBytes), was this motivated by a general discussion/ticket, or was it just to improve conformance to the coding standard by being explicit?
3.  There are a lot of good but seemingly unrelated changes in this patchset, were those necessary to fix the 9135/8985 or were other issues addressed that should be called out in a ticket or summary somewhere?

If you can look at this I will happily re-review, however, I don't think my green light would be enough to merge.  I strongly believe someone with deeper knowledge of twisted.web should review this to ensure there aren't lurking complications with some of the changes.

File: src/twisted/web/resource.py
Comment (rodrigc) [reply src/twisted/web/resource.py:247]: No, you are mistaken.  With Python 2.7, getattr() works with a unicode argument.
On Python 3, getattr() will fail with a bytes argument.  So it is safe to use getattr() with a unicode arg on Python 2.7 and Python 3.  See: https://github.com/python/cpython/blob/2.7/Objects/object.c#L1172

**Comment by rodrigc** [comment]
It's hard to pull out changes into separate tickets, because this code is interdependent.

File: src/twisted/web/test/requesthelper.py
Comment (rodrigc) [reply src/twisted/web/test/requesthelper.py:352]: There is not a lot of input validation done on all these arguments to begin with.
I added https://github.com/twisted/twisted/pull/911/commits/38ee5f3ca983c486521c9cffad5cb8afdbc3f54a which will fail on the testcase you provided.

**Comment by rodrigc** [comment]
The functions in `twisted.python.compat` were written a long time ago, during the initial work to port Twisted to Python 3.
They are OK, and got the ball rolling, but are not always the best.
After porting many Twisted modules to Python 3, I am more comfortable with being explicit
in some areas related to bytes/unicode encoding, and not depending on all the functions in twisted.python.compat.
Some of the functions such as networkString() are OK, but not the best.

**Comment by rodrigc** [comment]
Using a local checkout of Twisted from this branch, plus a local checkout of buildbot,
I used this script to checkout treq, klein, and builtbot into a Python 2.7 virtualenv:

```
#!/bin/sh

python2 -m virtualenv testit-2.7
. testit-2.7/bin/activate
python -m pip install --upgrade pip
python -m pip install --upgrade 'treq[dev,test]'
python -m pip install --upgrade 'klein'
cd $HOME/buildbot4
python -m pip install --upgrade -r requirements-ci.txt
python -m pip uninstall -y Twisted
cd $HOME/twisted6
python setup.py develop
```

and this script for a Python 3.6 virtualenv:

```
#!/bin/sh

python3 -m venv testit-3.6
. testit-3.6/bin/activate
python -m pip install --upgrade pip
python -m pip install --upgrade 'treq[dev,test]'
python -m pip install --upgrade 'klein'
cd $HOME/buildbot4
python -m pip install --upgrade -r requirements-ci.txt
python -m pip uninstall -y Twisted
cd $HOME/twisted6
python setup.py develop
```

In each virtualenv, I then ran these commands to test treq, klein, and buildbot against this Twisted branch:

```
trial treq
trial klein
trial buildbot.test
trial buildbot_worker.test
```

The tests passed.

File: src/twisted/web/test/requesthelper.py
Comment (jlitzingerdev) [reply src/twisted/web/test/requesthelper.py:352]: Thanks!  It was primarily the change in validation that drew my attention, not that I was expecting your changes to add new validation.

File: src/twisted/web/resource.py
Comment (jlitzingerdev) [reply src/twisted/web/resource.py:247]: So I am, thank you for the clarification!

**Comment by jlitzingerdev** [comment]
@rodrigc You have done a ton of work and testing with this PR, thank you!  My concerns have been addressed, did you have any thoughts on my request for a non misc newsfragment (I realize you haven't resubmitted for review, sorry if this is jumping ahead)?

As I'd like to help see this over the finish line I updated Twisted to this branch (in a sandbox) to test any impact to one of my use cases (HTTPS long polling on an embedded system) and its unit tests.  Everything worked as expected.  I have two others (TLS mutual auth and an HTTP server) I can test in a bit that can hopefully add some confidence.

**Comment by jlitzingerdev** [comment]
Why not introduce new compatibility functions as Glyph requested this in 3 in his review of 8985 (https://twistedmatrix.com/trac/ticket/8985#comment:15)?

It seems relevant for two reasons:

1.  Since this PR deviates from existing approaches to compatibility, it raises the question for non-committers --  which way is "correct"?  The documented, advertised compatibility module, or these deviations buried in source.  Odds are unless one is aware of this thread the former will be selected.  Does that mean proliferation of the "incorrect" way of doing things?   That seems bad.

2.  While there is no question there are improvements to what is in trunk, there are also changes well outside the scope of 8985 and 9135, with the goal of fixing bytes/str issues.  Given that, and the move away from existing compatibility functions, the introduction/deprecation seems directly applicable to what this PR is trying to achieve -- better bytes/str handling for Python 3.

**Comment by rodrigc** [comment]
Moving to https://github.com/twisted/twisted/pull/967

**Pull request closed without merging**
